src/include/storage/copier/node_copier.h

```diff
@@ -1,136 +1,235 @@
-        common::offset_t startOffset, uint64_t numBlocks,
-        std::unique_ptr<parquet::arrow::FileReader> parquetReader)
-        : NodeCopySharedState<T>{filePath, pkIndex, startOffset}, numBlocks{numBlocks},
-          parquetReader{std::move(parquetReader)} {};
-    std::unique_ptr<NodeCopyMorsel> getMorsel() override;
+    NodeCopier(std::shared_ptr<NodeCopySharedState> sharedState, HashIndexBuilder<T>* pkIndex,
+        const common::CopyDescription& copyDesc,
+        std::vector<std::unique_ptr<InMemNodeColumn>>& columns, common::column_id_t pkColumnID)
+        : sharedState{std::move(sharedState)}, pkIndex{pkIndex}, copyDesc{copyDesc},
+          columns{columns}, pkColumnID{pkColumnID} {
+        overflowCursors.resize(columns.size());
+    }
+    virtual ~NodeCopier() = default;
 
-public:
-    uint64_t numBlocks;
+    void execute(processor::ExecutionContext* executionContext);
 
-private:
-    std::unique_ptr<parquet::arrow::FileReader> parquetReader;
-};
+    inline virtual void finalize() { pkIndex->flush(); }
 
-class NodeCopier : public TableCopier {
-
-public:
-    NodeCopier(common::CopyDescription& copyDescription, std::string outputDirectory,
-        common::TaskScheduler& taskScheduler, catalog::Catalog& catalog, common::table_id_t tableID,
-        NodesStatisticsAndDeletedIDs* nodesStatisticsAndDeletedIDs)
-        : TableCopier{copyDescription, std::move(outputDirectory), taskScheduler, catalog, tableID,
-              nodesStatisticsAndDeletedIDs} {}
+    virtual std::unique_ptr<NodeCopier> clone() const {
+        return std::make_unique<NodeCopier>(sharedState, pkIndex, copyDesc, columns, pkColumnID);
+    }
 
 protected:
-    void initializeColumnsAndLists() override;
-
-    void populateColumnsAndLists(processor::ExecutionContext* executionContext) override;
-
-    void saveToFile() override;
+    virtual void executeInternal(std::unique_ptr<NodeCopyMorsel> morsel) {
+        throw std::runtime_error("Not implemented");
+    }
 
-    template<typename T>
+    static void copyArrayIntoColumnChunk(InMemColumnChunk* columnChunk, InMemNodeColumn* column,
+        arrow::Array& arrowArray, common::offset_t startNodeOffset,
+        common::CopyDescription& copyDescription, PageByteCursor& overflowCursor);
     static void populatePKIndex(InMemColumnChunk* chunk, InMemOverflowFile* overflowFile,
         common::NullMask* nullMask, HashIndexBuilder<T>* pkIndex, common::offset_t startOffset,
         uint64_t numValues);
+    static void appendPKIndex(InMemColumnChunk* chunk, InMemOverflowFile* overflowFile,
+        common::offset_t offset, HashIndexBuilder<T>* pkIndex) {
+        assert(false);
+    }
 
-    std::unordered_map<common::property_id_t, std::unique_ptr<InMemNodeColumn>> columns;
+protected:
+    std::shared_ptr<NodeCopySharedState> sharedState;
+    HashIndexBuilder<T>* pkIndex;
+    common::CopyDescription copyDesc;
+    std::vector<std::unique_ptr<InMemNodeColumn>>& columns;
```

Comment:
personally I usually use std::vector<InMemNodeColumn*> and not referencing the vector